Jackson BlackBird Support (for Java 11)

SudharakaP opened the pull request:
This adds ```jackson-blackbird``` support for java 11 as discussed in #9934. 

-   Please make sure the below checklist is followed for Pull Requests.

-   [x] [Travis tests](https://travis-ci.org/jhipster/generator-jhipster/pull_requests) are green
-   [x] Tests are added where necessary
-   [x] Documentation is added/updated where necessary
-   [x] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/master/CONTRIBUTING.md) are followed

<!--
Please also reference the issue number in a commit message to [automatically close the related Github issue](https://help.github.com/articles/closing-issues-via-commit-messages/)

Note: It is also possible to add `[skip ci]` to your commit message to skip Travis tests
-->

DanielFran commented on generators/server/templates/pom.xml.ejs:
```diff
@@ -206,10 +206,19 @@
+        <dependency>
+            <groupId>org.jdbi.jackson-contrib</groupId>
+            <artifactId>jackson-module-blackbird</artifactId>
+            <version>0.0.1</version>
```
Can you please create a property for the version: jackson-module-blackbird.version?

SudharakaP replied:
@DanielFran : Thanks; done 😄

cbornet commented on generators/app/prompts.js:
```diff
@@ -216,3 +217,50 @@ function askModulesToBeInstalled(done, generator) {
+
+function askJacksonBlackBirdToBeInstalled(done, generator) {
+    const PROMPT = {
+        type: 'input',
```
This should be a type "confirm" using a boolean

SudharakaP replied:
@cbornet : Thanks. Done 👍

cbornet commented on generators/app/prompts.js:
```diff
@@ -216,3 +217,50 @@ function askModulesToBeInstalled(done, generator) {
+    this.prompt(PROMPT).then(prompt => {
+        this.javaVersion = prompt.javaVersion;
+        if (prompt.javaVersion === '11') {
+            askJacksonBlackBirdToBeInstalled(done, this);
```
Shouldn't this be a filter in the blackbird "validate" prompt ?

SudharakaP replied:
@cbornet : Sorry, I am not sure I understand you here. You mean the the prompt for java version should go after the prompt for blackbird prompt?

cbornet replied:
Sorry I should have said in the "when" field of blackbird prompt, test the Java version (so after the Java version prompt).

SudharakaP replied:
@cbornet : Thanks for the clarification. Now understand what you mean; I haven't done any yeoman generator development and just now realized that ```Inquirer.js``` provides the ```when``` attribute. 

Refactored the code; let me know if you see any further issues. 😄

DanielFran commented:
@SudharakaP Gradle settings also require changes, see for example https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/build.gradle.ejs#L63

SudharakaP commented:
@DanielFran @cbornet : Thanks much for your comments and reviews. I'll fix all of these today evening or tomorrow morning. 😄

SudharakaP commented:
> @SudharakaP Gradle settings also require changes, see for example:
> 
> * https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/build.gradle.ejs#L63
> * https://github.com/jhipster/generator-jhipster/blob/master/generators/server/templates/build.gradle.ejs#L285
> 
> Note: I suggest to implement those changes on top of #10244 that should be merged soon... or wait for it.

@DanielFran : Thanks for pointing this out. Somehow I forgot about Gradle.

SudharakaP commented:
I've fixed some issues and also got all the tests pass at my end (I believe there was one test failing before). However two tests are failing in the azure pipelines and I don't think it's related to this. Could somebody please restart those two by any chance? 

@DanielFran : Added the Gradle stuff. Please let me know if you see anything else to be changed.

DanielFran commented on generators/server/templates/gradle.properties.ejs:
```diff
@@ -35,6 +35,7 @@ hibernate_version=5.3.10.Final
 mapstruct_version=1.3.0.Final
 
 liquibase_hibernate5_version=3.7
+jackson_module_blackbird_version=0.0.1
```
Please add condition <%_ if (javaVersion === '11' && useJacksonBlackBird) { _%>

SudharakaP replied:
Sorry, silly mistake; missed that somehow. Added it 😄

murdos commented on generators/app/prompts.js:
```diff
@@ -216,3 +218,43 @@ function askModulesToBeInstalled(done, generator) {
+        return;
+    }
+
+    const choices = [{ name: 'Java 11', value: '11' }, { name: 'Java 8', value: '1.8' }];
```
We also support Java 12 (and may also support Java 13 soon), so we should probably either reword things a bit differently ("Java >= 11") or include other supported java versions.

PierreBesson commented:
I thought we agreed to not add a new question for java 8 or 11 ! If it's only to add blackbird then I'd rather we wait until JH v7 where we will drop java 8 support.
ping @jdubois

SudharakaP commented:
@PierreBesson @murdos : One thing off the top of my head; ask the blackbird prompt if ```JAVA_HOME``` is set to Java11+. We can also set the ```java.version``` tag in pom.xml based on ```JAVA_HOME```. What do you think about this approach? I think this way we don't have to wait until JH7.

murdos commented:
> 
> 
> @PierreBesson @murdos : One thing off the top of my head; ask the blackbird prompt if `JAVA_HOME` is set to Java11+. We can also set the `java.version` tag in pom.xml based on `JAVA_HOME`. What do you think about this approach? I think this way we don't have to wait until JH7.

This approach is not correct, because you can generate a project for a different target than what you have on the machine that is used for generating the project.

SudharakaP commented:
> > @PierreBesson @murdos : One thing off the top of my head; ask the blackbird prompt if `JAVA_HOME` is set to Java11+. We can also set the `java.version` tag in pom.xml based on `JAVA_HOME`. What do you think about this approach? I think this way we don't have to wait until JH7.
> 
> This approach does not work, because you can generate a project for a different target than what you have on the machine that is used for generating the project.

@murdos : Ah, got it; didn't think about that. Thanks 👍 .

murdos commented:
>  
> I thought we agreed to not add a new question for java 8 or 11 ! If it's only to add blackbird then I'd rather we wait until JH v7 where we will drop java 8 support.

I vaguely remember this decision, but I have not been able to retrieve the reason of this choice. 

Since we now have a new java version released every 6 months, I think having the question could have some benefits:
- be able to be more specific in source and target versions specified in Maven and Gradle. For example if someone currently want to work with Java 11 language feature, they should manually change their configuration after the project has been generated
- ditto for generated CI stuff
- potentially generate different code and/or dependency (like here) according to the target java version
- explicitly show java versions that jhipster supports
- ...

SudharakaP commented:
> I vaguely remember this decision, but I have not been able to retrieve the reason of this choice.
> 
> Since we now have a new java version released every 6 months, I think having the question could have some benefits:
> 
> * be able to be more specific in source and target versions specified in Maven and Gradle. For example if someone currently want to work with Java 11 language feature, they should manually change their configuration after the project has been generated
> * ditto for generated CI stuff
> * potentially generate different code and/or dependency (like here) according to the target java version
> * explicitly show java versions that jhipster supports
> * ...

Exactly, we have a [jhipster based project](https://github.com/python-sinhala-education-society/OpenLearnr) and I had to go an change the pom file manually to update it to Java 11. So I do see having a prompt has benefits. 

@murdos : Anyways, I've updated the prompt as you specified. 

Feel free to let me know if there's any other changes you guys could see.

PierreBesson commented:
Yes I guess the additional prompt for java version could have some use but as version are released more and more frequently, I'd much rather that we support only one version officially in the future as it makes the CI build matrix increase exponentially !  And our CI is already too slow with too many possibilities to test.

Supporting java 8 was necessary because of the huge amount of breaking changes with java 9 but I don't think an upgrade like this will happen in the near future.

So I'm still against the question. I'm appealing to @jdubois, @pascalgrimaud, @deepu105 for the final decision on this.
